README.md
# grunt-sass-replace

> replaces sass values with grunt

[![NPM][16]][17]

## Getting Started
This plugin requires Grunt `~0.4.5`

If you haven't used [Grunt][1] before, be sure to check out the [Getting Started][2] guide, as it explains 
how to create a [Gruntfile][3] as well as install and use Grunt plugins. Once you're familiar with that process, 
you may install this plugin with this command:

```shell
npm install grunt-sass-replace --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-sass-replace');
```

## The 'sass-replace' task

### Overview
In your project's Gruntfile, add a section named `sass-replace` to the data object passed into `grunt.initConfig()`.

```js
grunt.initConfig({
  'sass-replace': {
    files: [
      {
        src: 'my-styles.scss',
        dest: 'my-new-styles.scss'
      }
    ],
    options: {
      variables: [
        {
          name: 'my-var',
          from: '10px',
          to: '3.333em'
        }
      ],
      imports: [
        {
          from: 'http://foo.wat.com',
          to: 'http://bar.wat.com'
        }
      ]
    }
  },
});
```

### Options

#### options.variables

_**Type:** `Array<Object>`_  
_**Default value:** none_  
_**Mandatory:** either this or [`options.imports`][4] must be set_  

A collection of [`VariableInstruction`][13]s to describe replacements of sass variable values.

##### VariableInstruction

_**Type:** `Object`_  
_**Default value:** none_  
_**Mandatory:** at least one `VariableInstruction` must be available on the [`options.variables`][8] collection_  

An object representing a single instruction in the `options.variables` collection.  
The [`to`][5] field is mandatory, and either [`from`][6], [`name`][7] or both can be used to filter the lookup.

When both `name` and `from` are used, they both apply as filters. As `name` can be passed as a string *or* as a regular 
expression, it presents two common use cases:

- passing both fields as strings can be used to replace values in duplicate variable declarations, e.g. a 
variable overriding a variable with the same name in another file.
- passing a string to `from` and a regular expression to `name` can be used to achieve finer filtering, matching
variables which have a certain value, by several names.

###### VariableInstruction.name

_**Type:** `String|RegExp`_  
_**Default value:** `/\S+/` (match at least one non-whitespace character)_  
_**Mandatory:** either this or the [`from`][6] field must be set_  

The variable name for lookup, without the leading `$`.
 
When a string is passed, the literal value is searched, e.g. `my-var` will match `$my-var: "foo"`.

A regular expression can be passed in its literal form or as a `RegExp` instance.  
e.g. `/my[-_]?[Vv]ar/` or `new RegExp('my[-_]?[Vv]ar')` will both match `$my-var: "foo"`, `$my_var: "foo"` and `$myVar: "foo"`.

When using a `RegExp` instance, only its source is used for the lookup (flags are ignored).

###### VariableInstruction.from

_**Type:** `String|Number|Boolean`_  
_**Default value:** none_  
_**Mandatory:** either this or the [`name`][7] field must be set_  

The variable's current value for lookup. Note that `!default` statements are not captured for replacement, and are 
kept as is ([post an issue][12] if you want to be able to alter `!default` statements as well).

See [note on using double quotes when replacing string values][9].

###### VariableInstruction.to

_**Type:** `String|Number|Boolean`_  
_**Default value:** none_  
_**Mandatory:** yes_  

A new value for the matched variable.

See [note on using double quotes when replacing string values][9].


##### Note on replacing string values

When replacing variable values of type string, e.g. `$my-var: "foo"` &rarr; `$my-var: "bar"`, the variable value's 
surrounding double quotes may be omitted from the `from`/`to` field's value, as they will be captured and passed on to 
the replacement.  
However, it is advised to prefer the explicit notation and always surround string values with double quotes, for better 
readability and to avoid confusion or unexpected behavior.

:+1:  
```javascript
  {
    from: '"foo"',
    to: '"bar"'
  }
  ...
```

:-1:  
```javascript
  {
    from: 'foo',
    to: 'bar'
  }
  ...
```


#### options.imports

_**Type:** `Array<Object>`_  
_**Default value:** none_  
_**Mandatory:** either this or [`options.variables`][8] must be set_  

A collection of [`ImportInstruction`][14]s to describe replacements of `@import` paths.

##### ImportInstruction

_**Type:** `Object`_  
_**Default value:** none_  
_**Mandatory:** at least one `ImportInstruction` must be available on the [`options.imports`][4] collection_  

An object representing a single instruction in the `options.imports` collection.  
Both [`from`][10] and [`to`][11] fields are mandatory.

###### ImportInstruction.from

_**Type:** `String`_  
_**Default value:** none_  
_**Mandatory:** yes_  

The import path(s) current value for lookup. Captures only the path contents, i.e. everything between the surrounding 
quotes, or inside a `url()`.

Capturing of everything after the `@import` (including quotes or `url()`s) is currently not supported, 
[post an issue][12] if you find it useful. Also, passing regular expressions was not tested, but probably works.

###### ImportInstruction.to

_**Type:** `String`_  
_**Default value:** none_  
_**Mandatory:** yes_  

A new value for the matched import path(s).


### Usage Examples

#### Replacing variables by current value

This example also shows the usage of various types in variable values.

```js
grunt.initConfig({
  'sass-replace': {
    files: {
     'dest/my-styles.scss': 'src/my-styles.scss'
    },
    options: {
      variables: [
        {
          from: '"foo"',
          to: '"bar"'
        },
        {
          from: 10,
          to: '10%'
        },
        {
          from: 3.333,
          to: 6.666
        },
        {
          from: '10px',
          to: '20em'
        }
      ]
    }
  }
});
```

#### Replacing variables by name

```js
grunt.initConfig({
  'sass-replace': {
    files: {
      'dest/my-styles.scss': 'src/my-styles.scss'
    },
    options: {
      variables: [
        {
          name: 'my-var',
          to: 'bar'
        }
      ]
    }
  }
});
```

#### Replacing variables by name and current value

```js
grunt.initConfig({
  'sass-replace': {
    files: {
      'dest/my-styles.scss': 'src/my-styles.scss'
    },
    options: {
      variables: [
        {
          name: 'my-var',
          from: '"foo"',
          to: '"bar"'
        }
      ]
    }
  }
});
```

#### Replacing variables using regular expressions

```js
grunt.initConfig({
  'sass-replace': {
    files: {
      'dest/my-styles.scss': 'src/my-styles.scss'
    },
    options: {
      variables: [
        {
          name: new RegExp('my[-_]?[Vv]ar'),
          from: '"foo"',
          to: 1000000000000
        },
        {
          name: /my[-_]?[Nn]um[-_]?[Vv]ar/,
          to: -1
        }
      ]
    }
  }
});
```

#### Replacing imports

```js
grunt.initConfig({
  'sass-replace': {
    files: {
      'dest/my-styles.scss': 'src/my-styles.scss'
    },
    options: {
      imports: [
        {
          from: 'foo',
          to: 'bar'
        },
        {
          from: 'foo.css',
          to: 'bar.css'
        },
        {
          from: 'foo.scss',
          to: 'bar.scss'
        },
        {
          from: 'http://wat.com/foo',
          to: 'http://wat.com/bar'
        },
        {
          from: 'http://wat.tha.fuck.com/foo',
          to: 'http://wat.tha.fuck.com/bar'
        },
        {
          from: 'http://wat.com/foo?family=#{$family}',
          to: 'http://wat.com/bar?family=#{$family}'
        },
        {
          from: 'foo-foo',
          to: 'bar-bar'
        },
        {
          from: 'foo-foo-foo',
          to: 'bar-bar-bar'
        }
      ]
    }
  }
});
```

## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed 
functionality, lint and test your code.

To allow testing of negative flows, the tests are run via npm, which is running the main task with the `--force` flag.
To run the tests, run:

```shell
npm run test
```

## Release History

See [the changelog][15].

## License

Copyright (c) 2016 Eliran Malka. Licensed under the WTFPL license.







[1]: http://gruntjs.com/
[2]: http://gruntjs.com/getting-started
[3]: http://gruntjs.com/sample-gruntfile
[4]: #optionsimports
[5]: #variableinstructionto
[6]: #variableinstructionfrom
[7]: #variableinstructionname
[8]: #optionsvariables
[9]: #note-on-replacing-string-values
[10]: #importinstructionfrom
[11]: #importinstructionto
[12]: https://github.com/eliranmal/grunt-sass-replace/issues
[13]: #variableinstruction
[14]: #importinstruction
[15]: CHANGELOG.md
[16]: https://img.shields.io/npm/v/grunt-sass-replace.svg?style=flat-square
[17]: https://www.npmjs.com/package/grunt-sass-replace


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.18"></a>
## [0.1.18](https://github.com/eliranmal/grunt-sass-replace/compare/v0.1.0...v0.1.18) (2016-09-29)


### Bug Fixes

* **task:** fix flawed validation ([6a252d8](https://github.com/eliranmal/grunt-sass-replace/commit/6a252d8))



<a name="0.1.0"></a>
# 0.1.0 (2016-09-28)


### Bug Fixes

* **task:** add validation for variables instruction from field allowed types ([a5eb607](https://github.com/eliranmal/grunt-sass-replace/commit/a5eb607))
* **task:** add validation for variables instruction to field allowed types ([9e6deb3](https://github.com/eliranmal/grunt-sass-replace/commit/9e6deb3))
* **task:** update task description ([7f7a493](https://github.com/eliranmal/grunt-sass-replace/commit/7f7a493))
* **tests:** fix expectation of variables name-to ([952b988](https://github.com/eliranmal/grunt-sass-replace/commit/952b988))


### Features

* **variables:** allow regex in "name" field ([12acab7](https://github.com/eliranmal/grunt-sass-replace/commit/12acab7))


test/fixtures/run.sh
# just a silly file for testing invalid files

test/sass-replace-test.js
'use strict';

var grunt = require('grunt');

/*
 ======== A Handy Little Nodeunit Reference ========
 https://github.com/caolan/nodeunit

 Test methods:
 test.expect(numAssertions)
 test.done()
 Test assertions:
 test.ok(value, [message])
 test.equal(actual, expected, [message])
 test.notEqual(actual, expected, [message])
 test.deepEqual(actual, expected, [message])
 test.notDeepEqual(actual, expected, [message])
 test.strictEqual(actual, expected, [message])
 test.notStrictEqual(actual, expected, [message])
 test.throws(block, [error], [message])
 test.doesNotThrow(block, [error], [message])
 test.ifError(value)
 */

exports['sass-replace'] = {

    setUp: function (done) {
        // setup here if necessary
        done();
    },

    variablesFromTo: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/variables/from-to.scss');
        var expected = grunt.file.read('test/expected/variables/from-to.scss');
        test.equal(actual, expected, 'should replace sass variable values, filtering by variable old value');

        test.done();
    },

    variablesNameTo: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/variables/name-to.scss');
        var expected = grunt.file.read('test/expected/variables/name-to.scss');
        test.equal(actual, expected, 'should replace sass variable values, filtering by variable name');

        test.done();
    },

    variablesNameFromTo: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/variables/name-from-to.scss');
        var expected = grunt.file.read('test/expected/variables/name-from-to.scss');
        test.equal(actual, expected, 'should replace sass variable values, filtering by variable name and old value');

        test.done();
    },

    variablesRegexNameTo: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/variables/regex-name-to.scss');
        var expected = grunt.file.read('test/expected/variables/regex-name-to.scss');
        test.equal(actual, expected, 'should replace sass variable values, filtering by variable name as regex');

        test.done();
    },

    variablesRegexNameFromTo: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/variables/regex-name-from-to.scss');
        var expected = grunt.file.read('test/expected/variables/regex-name-from-to.scss');
        test.equal(actual, expected, 'should replace sass variable values, filtering by variable name as regex, and by variable value');

        test.done();
    },

    importsFromTo: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/imports/from-to.scss');
        var expected = grunt.file.read('test/expected/imports/from-to.scss');
        test.equal(actual, expected, 'should replace sass import values, filtering by old value');

        test.done();
    },

    mixedVariables: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/mixed/variables.scss');
        var expected = grunt.file.read('test/expected/mixed/variables.scss');
        test.equal(actual, expected, 'should only replace sass variable values in a file with both imports and variables');

        test.done();
    },

    mixedImports: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/mixed/imports.scss');
        var expected = grunt.file.read('test/expected/mixed/imports.scss');
        test.equal(actual, expected, 'should only replace sass import values in a file with both imports and variables');

        test.done();
    },

    mixedAll: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/mixed/all.scss');
        var expected = grunt.file.read('test/expected/mixed/all.scss');
        test.equal(actual, expected, 'should replace sass import values and variable values in a file with both imports and variables');

        test.done();
    },

    invalidFileExtensionSrc: function (test) {
        test.expect(1);

        var isDestFileExists = grunt.file.exists('tmp/mixed/invalid-ext-src.scss');
        test.equal(isDestFileExists, false, 'should not write any files');

        test.done();
    },

    invalidFileExtensionDest: function (test) {
        test.expect(1);

        var isDestFileExists = grunt.file.exists('tmp/mixed/invalid-ext-dest.md');
        test.equal(isDestFileExists, false, 'should not write any files');

        test.done();
    },

    invalidFileExtensionSrcDest: function (test) {
        test.expect(1);

        var isDestFileExists = grunt.file.exists('tmp/mixed/invalid-ext-src-dest.md');
        test.equal(isDestFileExists, false, 'should not write any files');

        test.done();
    },

    missingOptions: function (test) {
        test.expect(1);

        var isDestFileExists = grunt.file.exists('tmp/mixed/no-options.scss');
        test.equal(isDestFileExists, false, 'should not write any files');

        test.done();
    }

};


test/sass_replace_test.js
'use strict';

var grunt = require('grunt');

/*
 ======== A Handy Little Nodeunit Reference ========
 https://github.com/caolan/nodeunit

 Test methods:
 test.expect(numAssertions)
 test.done()
 Test assertions:
 test.ok(value, [message])
 test.equal(actual, expected, [message])
 test.notEqual(actual, expected, [message])
 test.deepEqual(actual, expected, [message])
 test.notDeepEqual(actual, expected, [message])
 test.strictEqual(actual, expected, [message])
 test.notStrictEqual(actual, expected, [message])
 test.throws(block, [error], [message])
 test.doesNotThrow(block, [error], [message])
 test.ifError(value)
 */

exports.sass_replace = {
    setUp: function (done) {
        // setup here if necessary
        done();
    },
    default_options: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/default_options');
        var expected = grunt.file.read('test/expected/default_options');
        test.equal(actual, expected, 'should describe what the default behavior is.');

        test.done();
    },
    custom_options: function (test) {
        test.expect(1);

        var actual = grunt.file.read('tmp/custom_options');
        var expected = grunt.file.read('test/expected/custom_options');
        test.equal(actual, expected, 'should describe what the custom option(s) behavior is.');

        test.done();
    },
};


tasks/lib/util.js
/*
 * grunt-sass-replace
 * https://github.com/eliranmal/grunt-sass-replace
 *
 * Copyright (c) 2016 eliranmal
 * Licensed under the WTFPL license.
 */

var escapeStringRegexp = require('escape-string-regexp');

var util = module.exports = {

    isEmpty: function (val) {
        return !val || !util.hasProps(val);
    },

    hasProps: function (iterable) {
        for (var p in iterable) {
            return true;
        }
    },

    isScssFile: function (path) {
        return util.hasExtension(path, 'scss');
    },

    hasExtension: function (path, ext) {
        return util.endsWith(path, '.' + ext);
    },

    endsWith: function (str, term) {
        return str.indexOf(term) === str.length - term.length;
    },

    asRegexString: function (str) {
        return escapeStringRegexp(str + '');
    },

    stringify: function (json) {
        return JSON.stringify(json, function (key, val) {
            if (util.isRegex(val)) {
                return val.source;
            }
            return val;
        }, 2);
    },

    isString: function (val) {
        return typeof val === 'string';
    },

    isNumber: function (val) {
        return !isNaN(val) && typeof val === 'number';
    },

    isBoolean: function (val) {
        return typeof val === 'boolean';
    },

    isRegex: function (val) {
        return val instanceof RegExp;
    },

    isUndefined: function (val) {
        return typeof val === 'undefined';
    }

};

tasks/sass-replace.js
/*
 * grunt-sass-replace
 * https://github.com/eliranmal/grunt-sass-replace
 *
 * Copyright (c) 2016 eliranmal
 * Licensed under the WTFPL license.
 */

module.exports = function (grunt) {
    'use strict';

    var path = require('path'),
        util = require(path.resolve(__dirname, './lib/util')),
        sassReplace = require(path.resolve(__dirname, './lib/main')).init(grunt);

    // load 3rd party tasks
    grunt.task.loadTasks(path.resolve(__dirname, '../node_modules/grunt-string-replace/tasks'));

    grunt.registerMultiTask('sass-replace', 'replaces sass values', function () {
        var scssFiles,
            replacements,
            isScssFilesValid,
            isReplacementsValid,
            files = this.files,
            options = this.options();


        validate(files, 'no files passed.');

        if (util.isEmpty(options)) {
            grunt.fail.warn('no options passed.');
        } else if (util.isEmpty(options.variables) && util.isEmpty(options.imports)) {
            grunt.fail.warn('options must contain "variables", "imports", or both.');
        }

        scssFiles = files.filter(function (file) {
            var src = file.src,
                dest = file.dest,
                isSrcValid = !util.isEmpty(src) && src.filter(util.isScssFile).length,
                isDestValid = !util.isEmpty(dest) && util.isScssFile(dest);
            return isSrcValid && isDestValid;
        });

        isScssFilesValid = validate(scssFiles, 'no scss files found in passed files.');

        replacements = sassReplace.asStringReplacements(options);

        isReplacementsValid = validate(replacements, 'failed to resolve replacements.');

        if (isScssFilesValid && isReplacementsValid) {
            grunt.log.writeln('running string-replace task.');
            grunt.config.set('string-replace', {
                sass: {
                    files: scssFiles,
                    options: {
                        replacements: replacements
                    }
                }
            });
            grunt.task.run('string-replace:sass');
            grunt.log.ok().ok('sass-replace finished successfully.');
        }
    });

    function validate(value, invalidMessage, validMessage) {
        var isValid = !util.isEmpty(value);
        if (isValid) {
            if (validMessage) {
                grunt.log.ok(validMessage);
            }
        } else {
            if (invalidMessage) {
                grunt.fail.warn(invalidMessage);
            }
        }
        return isValid;
    }

};

tasks/sass_replace.js
/*
 * grunt-sass-replace
 * https://github.com/eliranmal/grunt-sass-replace
 *
 * Copyright (c) 2016 eliranmal
 * Licensed under the none license.
 */

'use strict';

module.exports = function (grunt) {

    // Please see the Grunt documentation for more information regarding task
    // creation: http://gruntjs.com/creating-tasks

    grunt.registerMultiTask('sass_replace', 'replaces sass declarations', function () {
        // Merge task-specific and/or target-specific options with these defaults.
        var options = this.options({
            punctuation: '.',
            separator: ', '
        });

        // Iterate over all specified file groups.
        this.files.forEach(function (f) {
            // Concat specified files.
            var src = f.src.filter(function (filepath) {
                // Warn on and remove invalid source files (if nonull was set).
                if (!grunt.file.exists(filepath)) {
                    grunt.log.warn('Source file "' + filepath + '" not found.');
                    return false;
                } else {
                    return true;
                }
            }).map(function (filepath) {
                // Read file source.
                return grunt.file.read(filepath);
            }).join(grunt.util.normalizelf(options.separator));

            // Handle options.
            src += options.punctuation;

            // Write the destination file.
            grunt.file.write(f.dest, src);

            // Print a success message.
            grunt.log.writeln('File "' + f.dest + '" created.');
        });
    });

};


Gruntfile.js
/*
 * grunt-sass-replace
 * https://github.com/eliranmal/grunt-sass-replace
 *
 * Copyright (c) 2016 eliranmal
 * Licensed under the WTFPL license.
 */

'use strict';

module.exports = function (grunt) {

    grunt.initConfig({

        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/**/*.js',
                '<%= nodeunit.tests %>'
            ],
            options: {
                jshintrc: '.jshintrc'
            }
        },

        clean: {
            tests: ['tmp']
        },

        // configuration to be run (and then tested).
        'sass-replace': {
            'variables_from-to': {
                files: [
                    {
                        //expand: true,
                        //flatten: true,
                        //src: ['test/fixtures/**/*'],
                        src: 'test/fixtures/variables.scss',
                        dest: 'tmp/variables/from-to.scss'
                    }
                ],
                options: {
                    variables: [
                        {
                            from: '"foo"',
                            to: '"bar"'
                        },
                        {
                            from: 10,
                            to: '10%'
                        },
                        {
                            from: 3.333,
                            to: 6.666
                        },
                        {
                            from: '10px',
                            to: '20em'
                        }
                    ]
                }
            },
            'variables_name-to': {
                files: [
                    {
                        src: 'test/fixtures/variables.scss',
                        dest: 'tmp/variables/name-to.scss'
                    }
                ],
                options: {
                    variables: [
                        {
                            name: 'my-var',
                            to: 'bar'
                        },
                        {
                            name: 'my-default-var',
                            to: 'bar'
                        },
                        {
                            name: 'my-num-var',
                            to: '20'
                        },
                        {
                            name: 'my-num-default-var',
                            to: 20
                        }
                    ]
                }
            },
            'variables_name-from-to': {
                files: [
                    {
                        src: 'test/fixtures/variables.scss',
                        dest: 'tmp/variables/name-from-to.scss'
                    }
                ],
                options: {
                    variables: [
                        {
                            name: 'my-var',
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            name: 'my-default-var',
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            name: 'my_num_var',
                            from: 3.333,
                            to: '6.666'
                        },
                        {
                            name: 'myNumDefaultVar',
                            from: '10px',
                            to: '20em'
                        }
                    ]
                }
            },
            'variables_regex-name-to': {
                files: {
                    'tmp/variables/regex-name-to.scss': 'test/fixtures/variables.scss'
                },
                options: {
                    variables: [
                        {
                            name: new RegExp('my[-_]?[Vv]ar'),
                            to: 1000000000000
                        },
                        {
                            name: /my[-_]?[Nn]um[-_]?[Vv]ar/,
                            to: -1
                        }
                    ]
                }
            },
            'variables_regex-name-from-to': {
                files: {
                    'tmp/variables/regex-name-from-to.scss': 'test/fixtures/variables.scss'
                },
                options: {
                    variables: [
                        {
                            name: new RegExp('my[-_]?[Vv]ar'),
                            from: 'foo',
                            to: 1000000000000
                        },
                        {
                            name: /my[-_]?[Nn]um[-_]?[Vv]ar/,
                            from: 10,
                            to: -1
                        }
                    ]
                }
            },
            'imports_from-to': {
                files: {
                    'tmp/imports/from-to.scss': 'test/fixtures/imports.scss'
                },
                options: {
                    imports: [
                        {
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            from: 'foo.css',
                            to: 'bar.css'
                        },
                        {
                            from: 'foo.scss',
                            to: 'bar.scss'
                        },
                        {
                            from: 'http://wat.com/foo',
                            to: 'http://wat.com/bar'
                        },
                        {
                            from: 'http://wat.tha.fuck.com/foo',
                            to: 'http://wat.tha.fuck.com/bar'
                        },
                        {
                            from: 'http://wat.com/foo?family=#{$family}',
                            to: 'http://wat.com/bar?family=#{$family}'
                        },
                        {
                            from: 'foo-foo',
                            to: 'bar-bar'
                        },
                        {
                            from: 'foo-foo-foo',
                            to: 'bar-bar-bar'
                        }
                    ]
                }
            },
            'mixed_imports': {
                files: {
                    'tmp/mixed/imports.scss': 'test/fixtures/mixed.scss'
                },
                options: {
                    imports: [
                        {
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            from: 'foo.css',
                            to: 'bar.css'
                        },
                        {
                            from: 'foo.scss',
                            to: 'bar.scss'
                        },
                        {
                            from: 'http://wat.com/foo',
                            to: 'http://wat.com/bar'
                        },
                        {
                            from: 'http://wat.tha.fuck.com/foo',
                            to: 'http://wat.tha.fuck.com/bar'
                        },
                        {
                            from: 'http://wat.com/foo?family=#{$family}',
                            to: 'http://wat.com/bar?family=#{$family}'
                        },
                        {
                            from: 'foo-foo',
                            to: 'bar-bar'
                        },
                        {
                            from: 'foo-foo-foo',
                            to: 'bar-bar-bar'
                        }
                    ]
                }
            },
            'mixed_variables': {
                files: {
                    'tmp/mixed/variables.scss': 'test/fixtures/mixed.scss'
                },
                options: {
                    variables: [
                        {
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            from: 10,
                            to: '10%'
                        },
                        {
                            from: '3.333',
                            to: 6.666
                        },
                        {
                            from: '10px',
                            to: '20em'
                        }
                    ]
                }
            },
            'mixed_all': {
                files: {
                    'tmp/mixed/all.scss': 'test/fixtures/mixed.scss'
                },
                options: {
                    variables: [
                        {
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            from: 10,
                            to: '10%'
                        },
                        {
                            from: '3.333',
                            to: 6.666
                        },
                        {
                            from: '10px',
                            to: '20em'
                        }
                    ],
                    imports: [
                        {
                            from: 'foo',
                            to: 'bar'
                        },
                        {
                            from: 'foo.css',
                            to: 'bar.css'
                        },
                        {
                            from: 'foo.scss',
                            to: 'bar.scss'
                        },
                        {
                            from: 'http://wat.com/foo',
                            to: 'http://wat.com/bar'
                        },
                        {
                            from: 'http://wat.tha.fuck.com/foo',
                            to: 'http://wat.tha.fuck.com/bar'
                        },
                        {
                            from: 'http://wat.com/foo?family=#{$family}',
                            to: 'http://wat.com/bar?family=#{$family}'
                        },
                        {
                            from: 'foo-foo',
                            to: 'bar-bar'
                        },
                        {
                            from: 'foo-foo-foo',
                            to: 'bar-bar-bar'
                        }
                    ]
                }
            },
            'invalid_file_ext_src': {
                files: {
                    'tmp/mixed/invalid-ext-src.scss': 'test/fixtures/doc.md'
                },
                options: {
                    variables: [
                        {
                            from: 'foo',
                            to: 'bar'
                        }
                    ]
                }
            },
            'invalid_file_ext_dest': {
                files: {
                    'tmp/mixed/invalid-ext-dest.md': 'test/fixtures/mixed.scss'
                },
                options: {
                    variables: [
                        {
                            from: 'foo',
                            to: 'bar'
                        }
                    ]
                }
            },
            'invalid_file_ext_src_dest': {
                files: {
                    'tmp/mixed/invalid-ext-src-dest.md': 'test/fixtures/doc.md'
                },
                options: {
                    variables: [
                        {
                            from: 'foo',
                            to: 'bar'
                        }
                    ]
                }
            },
            'missing_options': {
                files: {
                    'tmp/mixed/no-options.scss': 'test/fixtures/mixed.scss'
                }
            }
        },

        nodeunit: {
            tests: ['test/*-test.js']
        }

    });

    grunt.loadTasks('tasks');

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-nodeunit');

    grunt.registerTask('test', ['clean', 'sass-replace', 'nodeunit']);

    grunt.registerTask('default', ['jshint', 'test']);

};
